components/ConvertSteps/ConvertSteps.js

```diff
@@ -1,98 +1,202 @@
+
+      try {
+        updateStep(['setActive', stepIndex])
+        updateStep(['setStatus', stepIndex, STEP_WAITING])
+
+        // Awaiting confirmation
+        const transaction = await createTx()
 
-  const handleError = useCallback(() => {
-    setRetryingStep(null)
-    setStepperStatus(STEPPER_ERROR)
-  }, [])
+        onHashCreated && onHashCreated(transaction.hash)
+        updateStep(['setHash', stepIndex, transaction.hash])
 
-  const handleSuccess = useCallback(
-    optionalCb => {
-      return () => {
-        setRetryingStep(null)
-        optionalCb && optionalCb()
+        // Mining transaction
+        updateStep(['setStatus', stepIndex, STEP_WORKING])
+        await transaction.wait()
 
-        // Activate next step if not last
+        // Success
+        updateStep(['setStatus', stepIndex, STEP_SUCCESS])
+
+        // Activate next step or show as completed
         if (stepperStage < steps.length - 1) {
           setStepperStage(stepperStage + 1)
-        }
-
-        // Show overall stepper success if final step succeeds
-        if (stepperStage === steps.length - 1) {
+        } else {
           setStepperStatus(STEPPER_SUCCESS)
         }
+      } catch (err) {
+        // If there's a problem mining the transaction we catch it
+        // and visually feedback to the user
+        updateStep(['setStatus', stepIndex, STEP_ERROR])
+        setStepperStatus(STEPPER_ERROR)
+        console.error(err)
       }
     },
-    [stepperStage, steps]
+    [steps, stepperStage]
   )
 
+  const handleRetrySigning = useCallback(() => {
+    setStepperStatus(STEPPER_IN_PROGRESS)
+    attemptStepSigning(stepperStage)
+  }, [attemptStepSigning, stepperStage])
+
+  useEffect(() => {
+    attemptStepSigning(stepperStage)
+  }, [stepperStage, attemptStepSigning])
+
   const renderSteps = () => {
-    return steps.map((step, index) => (
-      <li
-        css={`
-          display: flex;
-        `}
-        key={index}
-      >
-        <ManageStep
-          key={index}
+    return steps.map((step, index) => {
+      const stepCurrentlyActive = stepperStage === steps.indexOf(step)
+
+      const renderStep = (
+        <Step
           title={step[0]}
           number={index + 1}
-          canRetry={retryingStep === index}
-          handleTx={step[1].createTx}
-          active={stepperStage === steps.indexOf(step)}
-          onHashCreation={step[1].hashCreated}
-          onSuccess={handleSuccess(step[1].success)}
-          onError={handleError}
+          active={stepState[index].active}
+          status={stepState[index].status}
+          transactionHash={stepState[index].hash}
         />
+      )
+
+      return (
+        <li
+          key={index}
+          css={`
+            display: flex;
+          `}
+        >
+          {/* Only render currently active step without divider on small screens */}
+          {layoutName === 'small' && stepCurrentlyActive && <>{renderStep}</>}
```

Comment:
It feels a bit odd to do a `map()` to return a single item, when we could break earlier before the `map()` to return the single item.

Perhaps we can break this down so the `renderSteps()` method calls a `renderStep()` method, which would return both the `<li>` and `<Step>`?

Something like:

```js
renderStep(step) {
  return (
    <li>
      <Step />
    </li>
  )
}

renderSteps() {
  return steps.map(step => (
    <>
      {renderStep(step)}
      {divider}
    <>
  ))
}
```

Reply:
Yes, this makes more sense. I've separated them and now perform the layout check inline within the main body of the template.